### PR TITLE
Migrate to Swift 4

### PR DESCRIPTION
**Pull request by jeffreylyg**

### CODE REVIEW

**Comment by jwhitley** [comment]
FYI, `.travis.yml` is still using [`osx_image: xcode8.3`](https://github.com/inamiy/RxAutomaton/blob/swift/3.0/.travis.yml#L6), so its tests are bound to fail on an Xcode 9/Swift 4 branch.

**Comment by jwhitley** [comment]
RxAutomaton.podspec needs to be updated to RxSwift 4.0, e.g.
> `  s.dependency "RxSwift", "~> 4.0"`

File: Demo/ViewController.swift [hunk 1]
```diff
@@ -38,7 +38,7 @@ class AutomatonViewController: UIViewController
         {
             return Observable<Int>.interval(interval, scheduler: MainScheduler.instance)
                 .take(count)
-                .scan(0) { $0.0 + 1 }
+                .scan(0) { ($0, $1).0 + 1 }
```
Comment (inamiy) [inline Demo/ViewController.swift]: `{ x, _ in x + 1 }` looks better.

File: Sources/Automaton.swift [hunk 2]
```diff
@@ -137,9 +137,9 @@ public final class Automaton<State, Input>
 
 // MARK: Private
 
-private func _compose<A, B, C>(_ g: @escaping (B) -> C, _ f: @escaping (A) -> B) -> (A) -> C
+private func _compose<A, B, C, D>(_ g: @escaping ((B) -> C), _ f: @escaping ((A, D) -> B)) -> ((A, D) -> C)
```
Comment (inamiy) [inline Sources/Automaton.swift]: Please change the use of type parameters in alphabetical order for readability, i.e.:

```swift
private func _compose<A, B, C, D>(_ g: @escaping ((C) -> D), _ f: @escaping ((A, B) -> C)) -> ((A, B) -> D)
```

**Review by inamiy — COMMENTED** [review]
@jeffreylyg 
Hi, thank you for fixes!
Overall looks good to me 👍 
Would you mind changing the parts I commented, and also the followings:

- Change target branch to `swift/4.0`
- Update `.swift-version`
- Update `Package.swift` and `Package.resolved`
- (If CI still fails) Comment-out `pod lib lint` in `.travis.yml`

**Review by inamiy — APPROVED** [review]
Thanks! Now it LGTM 👍

**Pull request merged**